### DISCUSSION TITLE
far:edit

### DISCUSSION THREAD

**Discussion by deep-soft** [elfmz/far2l#1832] [Ideas]
I tested last commits:
far:config
and
far:about
and they are very usefull.

Can be added other two commands for commandline:
far:edit
and
far:view
or better
edit:
and
view:
as in Far2 for Windows?

I searched in code and found references but only comments in:
far2l\src\fnparce.cpp
far2l\src\usermenu.cpp

And also can be added an option to not show panels after any valid command but after keypress?!
Like after an invalid/non existing command?!

Thank you for your great work!

**Comment by akruphi** [comment]
Good idea. Partially done in PR #1864

**Reply by deep-soft** [reply]
PR https://github.com/elfmz/far2l/pull/1864 tested and it's working!
Based on that I added short version of far commands:
`edit:` and `view:`
with some extra code
https://github.com/elfmz/far2l/pull/1864#issuecomment-1737816817

**Comment by akruphi** [comment]
Now realization `edit:` & `view:` not compatible with escaping by `\` which use in **Ctrl-Enter**/**Ctrl-F** for filenames with spaces and all `Opt.strQuotedSymbols` symbols.
@deep-soft is this problem for you?
I did not find by quick search in far2l inverse function for functions `EscapeSpace()` & `escapeSpace()` from `far2l/src/mix/strmix.cpp`

**Reply by deep-soft** [reply]
You're right; didn't observed this because every file I create does not contain spaces. But there are some people which use spaces a lot.
Not a problem right now, F3/F4 working for this filenames with spaces in name.

**Comment by akruphi** [comment]
@deep-soft please tell me why using commands `edit:` or `view:` is preferable to you than using keys **F4** and **F3**. If I  understand you really usecase I can make special command more consciously.

**Comment by deep-soft** [comment]
next step is using as in Far2 for Windows
edit:<ls -lR
or
edit:<fd | as-tree

I mean redirect command execution directly to Editor.

Far 2
https://github.com/FarGroup/FarManager/blob/3cf510210d6b1b8fe163729375c8297ebbb839c5/plugins/farcmds/FARCmds.cpp#L330

Far 3
https://github.com/FarGroup/FarManager/blob/6786705415cab812e4906ab7f887513569c99c85/plugins/farcmds/OpenCmd.cpp#L1050

**Reply by akruphi** [reply]
I don't know how in far2l grab stdout only from last command. This question only for @elfmz.

Now in far2l you can do command, then toggle off panels by **Ctrl+O** and obtain all terminal output in Editor by **F4** or Viewer by **F3**. Also before command you can without panels do **F8** to clear terminal output.

**Reply by deep-soft** [reply]
It's pretty complex; Far3 have this plugin, FarCmds, which defines far:view (view:), far:edit (edit:), ..., OS commands including redirection of outputs (StdOut and StdErr) to far editor.
https://github.com/FarGroup/FarManager/blob/6786705415cab812e4906ab7f887513569c99c85/plugins/farcmds/OpenCmd.cpp#L533

**Reply by deep-soft** [reply]
> I don't know how in far2l grab stdout only from last command. This question only for @elfmz.
> 
> Now in far2l you can do command, then toggle off panels by **Ctrl+O** and obtain all terminal output in Editor by **F4** or Viewer by **F3**. Also before command you can without panels do **F8** to clear terminal output.

This feature (redirect from stdout) is working now!
https://github.com/elfmz/far2l/pull/2361
Thank you

**Comment by deep-soft** [comment]
> Now in far2l you can do command, then toggle off panels by **Ctrl+O** and obtain all terminal output in Editor by **F4** or Viewer by **F3**. Also before command you can without panels do **F8** to clear terminal output.

Yes I know that (using Far from year 2000, and Far2l ~2years) and are some other methods to redirect stdout and stderr to file and after that open in far2l (via far2l /e or far2l /v) but the method used in Windows is more useful in some cases (needs less typing).

Thank you, @akruphi, and @elfmz for implementing this far commands (about, config, edit and view).

**Comment by nikk106** [comment]
Hello,
In Midnight Commander (MC), when editing a text file and needing to copy a long command to the terminal, pressing Ctrl+O temporarily hides the editor and shows a clean terminal. This makes copying and running commands very convenient.

In far2l, this workflow is less smooth — switching from the editor to the terminal while copying text is not as straightforward.
It would be helpful if far2l supported a similar feature, allowing quick terminal access from the editor without closing or minimizing it.
Thanks!

**Reply by akruphi** [reply]
> In far2l, this workflow is less smooth

Now far2l use windows far traditional and in Editor/Viewer **Ctrl**+**O** show current frozen build-in terminal screen, in which you can use **Alt**+**Ins** to grab any element from screen.

If you need real terminal you can by **Ctrl**+**Tab** or **F12** switch to panel and only in panel **Ctrl**+**O** to open build-in terminal.

Info: now the command line in far is not part of terminal, but it is inherent part of files panels.

**Comment by nikk106** [comment]
Thank you for your reply. F12 really helps. Good to know :)
